File: Emulator/src/main/content/global/skill/summoning/familiar/npc/SpiritCobraNPC.java
package content.global.skill.summoning.familiar.npc;

import content.global.skill.summoning.familiar.Familiar;
import content.global.skill.summoning.familiar.FamiliarSpecial;
import core.game.node.entity.combat.equipment.WeaponInterface;
import core.game.node.entity.player.Player;
import core.game.node.item.Item;
import core.plugin.Initializable;

/**
 * The type Spirit cobra npc.
 */
@Initializable
public class SpiritCobraNPC extends content.global.skill.summoning.familiar.Familiar {

    /**
     * Instantiates a new Spirit cobra npc.
     */
    public SpiritCobraNPC() {
        this(null, 6802);
    }

    /**
     * Instantiates a new Spirit cobra npc.
     *
     * @param owner the owner
     * @param id    the id
     */
    public SpiritCobraNPC(Player owner, int id) {
        super(owner, id, 5600, 12015, 3, WeaponInterface.STYLE_ACCURATE);
    }

    @Override
    public Familiar construct(Player owner, int id) {
        return new SpiritCobraNPC(owner, id);
    }

    @Override
    protected boolean specialMove(FamiliarSpecial special) {
        final Item item = (Item) special.getNode();
        final Egg egg = Egg.forEgg(item);
        if (egg == null) {
            owner.getPacketDispatch().sendMessage("You can't use the special move on this item.");
            return false;
        }
        owner.getInventory().replace(egg.getProduct(), item.getSlot());
        return true;
    }

    @Override
    public int[] getIds() {
        return new int[]{6802, 6803};
    }

    /**
     * The enum Egg.
     */
    public enum Egg {
        /**
         * The Cockatrice.
         */
        COCKATRICE(new Item(1944), new Item(12109)),
        /**
         * The Saratrice.
         */
        SARATRICE(new Item(5077), new Item(12113)),
        /**
         * The Zamatrice.
         */
        ZAMATRICE(new Item(5076), new Item(12115)),
        /**
         * The Guthatrice.
         */
        GUTHATRICE(new Item(5078), new Item(12111)),
        /**
         * The Coracatrice.
         */
        CORACATRICE(new Item(11964), new Item(12119)),
        /**
         * The Pengatrice.
         */
        PENGATRICE(new Item(12483), new Item(12117)),
        /**
         * The Vulatrice.
         */
        VULATRICE(new Item(11965), new Item(12121));

        private final Item egg;

        private final Item product;

        private Egg(Item egg, Item product) {
            this.egg = egg;
            this.product = product;
        }

        /**
         * For egg egg.
         *
         * @param item the item
         * @return the egg
         */
        public static Egg forEgg(final Item item) {
            for (Egg egg : values()) {
                if (egg.getEgg().getId() == item.getId()) {
                    return egg;
                }
            }
            return null;
        }

        /**
         * Gets egg.
         *
         * @return the egg
         */
        public Item getEgg() {
            return egg;
        }

        /**
         * Gets product.
         *
         * @return the product
         */
        public Item getProduct() {
            return product;
        }

    }
}

File: Emulator/src/main/content/global/skill/summoning/familiar/npc/SpiritCobraNPC.kt
package content.global.skill.summoning.familiar.npc

import content.global.skill.summoning.familiar.Familiar
import content.global.skill.summoning.familiar.FamiliarSpecial
import core.api.sendMessage
import core.game.node.entity.combat.equipment.WeaponInterface
import core.game.node.entity.player.Player
import core.game.node.item.Item
import core.plugin.Initializable
import org.rs.consts.Items
import org.rs.consts.NPCs

/**
 * The type Spirit cobra npc.
 */
@Initializable
class SpiritCobraNPC
/**
 * Instantiates a new Spirit cobra npc.
 *
 * @param owner the owner
 * @param id    the id
 */
/**
 * Instantiates a new Spirit cobra npc.
 */
@JvmOverloads constructor(owner: Player? = null, id: Int = NPCs.SPIRIT_COBRA_6802) :
    Familiar(owner, id, 5600, Items.SPIRIT_COBRA_POUCH_12015, 3, WeaponInterface.STYLE_ACCURATE) {
    override fun construct(owner: Player, id: Int): Familiar {
        return SpiritCobraNPC(owner, id)
    }

    override fun specialMove(special: FamiliarSpecial): Boolean {
        val item = special.node as Item
        val egg = Egg.forEgg(item)
        if (egg == null) {
            sendMessage(owner, "You can't use the special move on this item.")
            return false
        }
        owner.inventory.replace(egg.product, item.slot)
        return true
    }

    override fun getIds(): IntArray {
        return intArrayOf(NPCs.SPIRIT_COBRA_6802, NPCs.SPIRIT_COBRA_6803)
    }

    /**
     * The enum Egg.
     */
    enum class Egg(
        /**
         * Gets egg.
         *
         * @return the egg
         */
        val egg: Item,
        /**
         * Gets product.
         *
         * @return the product
         */
        val product: Item
    ) {
        /**
         * The Cockatrice.
         */
        COCKATRICE(Item(Items.EGG_1944), Item(Items.COCKATRICE_EGG_12109)),

        /**
         * The Saratrice.
         */
        SARATRICE(Item(Items.BIRDS_EGG_5077), Item(Items.SARATRICE_EGG_12113)),

        /**
         * The Zamatrice.
         */
        ZAMATRICE(Item(Items.BIRDS_EGG_5076), Item(Items.ZAMATRICE_EGG_12115)),

        /**
         * The Guthatrice.
         */
        GUTHATRICE(Item(Items.BIRDS_EGG_5078), Item(Items.GUTHATRICE_EGG_12111)),

        /**
         * The Coracatrice.
         */
        CORACATRICE(Item(Items.RAVEN_EGG_11964), Item(Items.CORAXATRICE_EGG_12119)),

        /**
         * The Pengatrice.
         */
        PENGATRICE(Item(Items.PENGUIN_EGG_12483), Item(Items.PENGATRICE_EGG_12117)),

        /**
         * The Vulatrice.
         */
        VULATRICE(Item(Items.VULTURE_EGG_11965), Item(Items.VULATRICE_EGG_12121));

        companion object {
            /**
             * For egg egg.
             *
             * @param item the item
             * @return the egg
             */
            fun forEgg(item: Item): Egg? {
                for (egg in values()) {
                    if (egg.egg.id == item.id) {
                        return egg
                    }
                }
                return null
            }
        }
    }
}
